### ISSUE TITLE
Event callback for client/server service?

### ISSUE THREAD

**Issue by jtschea** [continental/ecal#420]
Dear developers,

thanks a lot for this excellent product :smiley: 

For classes [CSubscriber](https://continental.github.io/ecal/_api/classeCAL_1_1CSubscriber.html#) and [CPublisher](https://continental.github.io/ecal/_api/classeCAL_1_1CPublisher.html) there's a method `AddEventCallback` to get notified when connection state has changed.

Is there something similar for both [CServiceClient](https://continental.github.io/ecal/_api/classeCAL_1_1CServiceClient.html) and [CServiceServer](https://continental.github.io/ecal/_api/classeCAL_1_1CServiceServer.html)?

There's the chance that a RPC call got lost if the server is not yet up and running and we wouldn't want to rely on `eCallState` for checking connection state.

Do you maybe have a suggestion for a workaround?

Thanks for your help and best regards,
Jupp

**Comment by rex-schilasky** [comment]
Hi Jupp,

thank you for the nice feedback. Currently we do not have an `AddEventCallback` method similar to `CSubscriber` and `CPublisher` for the services. There was no user request until today :).

But in general internally service server and client both do have methods that are called at least when a new matching service connects. So for at least a the events `service_server_connected` and `service_client_connected` could be implemented easily.

Let me check that and implement a first draft in a branch for reviewing.

Best, ReX

**Comment by rex-schilasky** [comment]
As a workaround there is another API method in `eCAL::Process`. With `eCAL::Process::AddRegistrationCallback` you can register a method that informs you if a new publisher, subscriber, service or process is registered.

```cpp
#include <ecal/ecal.h>
#include "ecal/pb/ecal.pb.h"

#include <iostream>

void OnServiceRegistration(const char* sample_, int sample_size_)
{
  eCAL::pb::Sample sample;
  if (sample.ParseFromArray(sample_, sample_size_))
  {
    std::cout << "service host : " << sample.service().hname() << std::endl;  // host name
    std::cout << "service uname: " << sample.service().uname() << std::endl;  // process name
    std::cout << "service name: "  << sample.service().sname() << std::endl << std::endl;  // service name
  }
}

int main(int argc, char **argv)
{
  // initialize eCAL API
  eCAL::Initialize(argc, argv, "monitoring registrations", eCAL::Init::None);

  // add service register callback function
  eCAL::Process::AddRegistrationCallback(reg_event_service, std::bind(OnServiceRegistration, std::placeholders::_1, std::placeholders::_2));

  while(eCAL::Ok()) eCAL::Process::SleepMS(100);

  // finalize eCAL API
  eCAL::Finalize();

  return(0);
}
```

See also [samples/cpp/monitoring/monitoring_reg/src/monitoring_reg.cpp](https://github.com/continental/ecal/blob/fc6891e9df37bfe95a4bb3d26214066a14a0f9c1/samples/cpp/monitoring/monitoring_reg/src/monitoring_reg.cpp)

**Comment by jtschea** [comment]
Dear ReX,

thanks a lot for both accepting the feature request and the provision of a workaround.

I will give it a try tomorrow and give feedback here soon.

Best regards,
Jupp

**Comment by jtschea** [comment]
Dear ReX,

sorry for answering late.

I tried out your workaround using `eCAL::Process::AddRegistrationCallback` and it worked like a charm. Thanks a lot :smiley: 

Do we still want to keep this issue open until the draft branch has been created? Maybe I can contribute a little by testing the feature.

Thanks again and best regards,
Jupp

**Comment by rex-schilasky** [comment]
Dear Jupp,

today finally I made a PR #427 for the new eCAL services API. You now can add event callbacks for services and clients. The cpp samples [math_server](https://github.com/continental/ecal/blob/a96695e2283a92b8958eb0337d57ca57495d4da7/samples/cpp/services/math_server/src/math_server.cpp) and [math_client](https://github.com/continental/ecal/blob/a96695e2283a92b8958eb0337d57ca57495d4da7/samples/cpp/services/math_client/src/math_client.cpp) demonstrate the use. The connect events will fire immediately the disconnect events after a timeout (between 1 and 5 seconds - still in work for improvement). 

Best, ReX

Here a "how to use" snippet ..

```cpp
#include <ecal/ecal.h>
#include <ecal/msg/protobuf/client.h>

#include <iostream>
#include <chrono>
#include <thread>

#include "math.pb.h"

// client state callback
void OnClientState(const eCAL::SClientEventCallbackData* data_, int* refcnt_)
{
  assert(data_);
  assert(refcnt_);

  switch (data_->type)
  {
  case client_event_connected:
    std::cout << "---------------------------------" << std::endl;
    std::cout << "Client connected to a service    " << std::endl;
    std::cout << "---------------------------------" << std::endl;
    (*refcnt_)++;
    break;
  case client_event_disconnected:
    std::cout << "-----------------------------------" << std::endl;
    std::cout << "Client disconnected from a service " << std::endl;
    std::cout << "-----------------------------------" << std::endl;
    (*refcnt_)--;
    break;
  default:
    std::cout << "Unknown client event." << std::endl;
    break;
  }

  std::cout << "Server Hostname : " << data_->attr.hname    << std::endl;
  std::cout << "Server Name     : " << data_->attr.sname    << std::endl;
  std::cout << "Server Process  : " << data_->attr.pname    << std::endl;
  std::cout << "Server PID      : " << data_->attr.pid      << std::endl;
  std::cout << "Server TCP Port : " << data_->attr.tcp_port << std::endl;
  std::cout << "---------------------------------"          << std::endl << std::endl;
}

// main entry
int main(int argc, char **argv)
{
  // initialize eCAL API
  eCAL::Initialize(argc, argv, "math client");

  // create math service client
  eCAL::protobuf::CServiceClient<MathService> math_service;

  // add event callbacks
  int refcnt(0);
  math_service.AddEventCallback(client_event_connected,    std::bind(OnClientState, std::placeholders::_2, &refcnt));
  math_service.AddEventCallback(client_event_disconnected, std::bind(OnClientState, std::placeholders::_2, &refcnt));

  while(eCAL::Ok())
  {
    // we are connected ..
    if (refcnt > 0)
    {
      // ... do your work here
    }
    // we are not connected
    else
    {
      std::cout << "Waiting for the service .." << std::endl << std::endl;
    }

    // sleep a second
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
  }

  // finalize eCAL API
  eCAL::Finalize();
  
  return(0);
}
```

**Comment by rex-schilasky** [comment]
This issue will be fixed with #372 and is part of PR #427
